### PR TITLE
Harden velocypack validator after fuzz testing

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/15649

Harden velocypack validator after fuzz testing.
Changes are from upstream https://github.com/arangodb/velocypack/pull/111

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports
  - [x] Backport for 3.9: this PR
  - [x] Backport for 3.8: https://github.com/arangodb/arangodb/pull/15662
  - [x] Backport for 3.7: https://github.com/arangodb/arangodb/pull/15663

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: 3rdParty/velocypack/src/Parser.cpp [hunk 1]
```diff
@@ -118,6 +118,17 @@ int Parser::skipWhiteSpace(char const* err) {
   throw Exception(Exception::ParseError, err);
 }
 
+void Parser::increaseNesting() {
```
Comment (cpjulia) [inline 3rdParty/velocypack/src/Parser.cpp:121]: would we throw exception, not just leave nesting as it is if can't increase it more?

**Review by cpjulia — APPROVED** [review]
LGTM

File: 3rdParty/velocypack/src/Validator.cpp [hunk 2]
```diff
@@ -121,14 +129,18 @@ bool Validator::validate(uint8_t const* ptr, std::size_t length, bool isSubPart)
     }
 
     case ValueType::Array: {
-      ++_level;
+      if (++_level >= options->nestingLimit) {
```
Comment (mchacki) [inline 3rdParty/velocypack/src/Validator.cpp:132]: The same comment as above applies here.

File: 3rdParty/velocypack/src/Validator.cpp [hunk 3]
```diff
@@ -121,14 +129,18 @@ bool Validator::validate(uint8_t const* ptr, std::size_t length, bool isSubPart)
...
+        throw Exception(Exception::TooDeepNesting);
```
Comment (mchacki) [inline 3rdParty/velocypack/src/Validator.cpp:133]: Is there any chance we can improve this error messages?
When debugging such an issue any additional context will help.
To reach this error the document to test typically is more complex, so e.g. char position or attributePath where the validator fails will be very helpful.

File: 3rdParty/velocypack/src/Validator.cpp [hunk 4]
```diff
@@ -220,7 +216,35 @@ bool Validator::validate(uint8_t const* ptr, std::size_t length, bool isSubPart)
...
+      length -= 2;
+    } else if (head == 0xef) {
+      // 8 bytes tag type
+      // the actual Slice (without tag) must be at least one byte long
+      validateBufferLength(1 + 8 + 1, length, true);
+      VELOCYPACK_ASSERT(length > 9);
+      ptr += 9;
+      length -= 9;
+    } else {
+      throw Exception(Exception::NotImplemented);
```
Comment (mchacki) [inline 3rdParty/velocypack/src/Validator.cpp:241]: Here again, what is not implemented would be a helpful error message addition for debugging

**Review by mchacki — APPROVED** [review]
In general LGTM.
I have left some suggestions for further improvements.

**Pull request merged**